round-based/src/state_machine/shared_state.rs

```diff
@@ -0,0 +1,337 @@
+
+struct SharedState<M> {
+    incoming_msg: Option<crate::Incoming<M>>,
+    outgoing_msg: Option<crate::Outgoing<M>>,
+    wants_recv_msg: bool,
+    wants_send_msg: bool,
+    yielded: bool,
+}
+
+impl<M> SharedState<M> {
+    pub fn new() -> Self {
+        Self {
+            incoming_msg: None,
+            outgoing_msg: None,
+            wants_recv_msg: false,
+            wants_send_msg: false,
+            yielded: false,
+        }
+    }
+}
+
+impl<M> SharedStateRef<M> {
+    pub fn new() -> Self {
+        Self(alloc::rc::Rc::new(core::cell::RefCell::new(
+            SharedState::new(),
+        )))
+    }
+
+    /// Any protocol-initated work (like flushing message to be sent, receiving message, etc.) can
+    /// only be scheduled when there was no other task scheduled.
+    ///
+    /// This method checks whether a task can be scheduled, and returns [`CanSchedule`] which
+    /// then can be used to schedule one task. If some other task is already scheduled, it returns
+    /// `Poll::Pending`.
+    pub fn can_schedule(&self) -> core::task::Poll<CanSchedule<&Self>> {
+        let s = self.0.borrow();
+        let can_poll = !s.wants_recv_msg && !s.wants_send_msg && !s.yielded;
+
+        if can_poll {
+            core::task::Poll::Ready(CanSchedule(self))
+        } else {
+            core::task::Poll::Pending
+        }
+    }
+
+    /// Puts a message to be sent into sending slot
+    ///
+    /// Returns `Err(msg)` if slot is already occupied (which means that flushing must be done
+    /// before sending a message).
+    ///
+    /// Note that it does not schedules that message to be sent. To do so, you need to schedule
+    /// flushing via [`CanSchedule::protocol_flushes_outgoing_msg`]
+    pub fn protocol_saves_msg_to_be_sent(
+        &self,
+        msg: crate::Outgoing<M>,
+    ) -> Result<(), crate::Outgoing<M>> {
+        let mut s = self.0.borrow_mut();
+        if s.outgoing_msg.is_some() {
+            return Err(msg);
+        }
+        s.outgoing_msg = Some(msg);
+        Ok(())
+    }
+
+    /// Takes outgoing message to be sent, but only if flushing was requested
+    pub fn executor_takes_outgoing_msg(&self) -> Option<crate::Outgoing<M>> {
+        let mut s = self.0.borrow_mut();
+        if s.wants_send_msg {
+            debug_assert!(s.outgoing_msg.is_some());
+            s.wants_send_msg = false;
+            s.outgoing_msg.take()
+        } else {
+            None
+        }
+    }
+
+    /// Checks if protocol has indicated that it waits for a new message from other parties.
+    /// Does not mutate the state
+    pub fn protocol_wants_more_messages(&self) -> bool {
+        let s = self.0.borrow();
+        s.wants_recv_msg
+    }
+
+    /// Checks if protocol yielded. Sets `yield_now = false`
+    pub fn executor_reads_and_resets_yielded_flag(&self) -> bool {
+        let mut s = self.0.borrow_mut();
+        let y = s.yielded;
+        s.yielded = false;
+        y
```

Comment:
`std::mem::replace(&mut s.yielded, false)`. But it really doesn't matter, feel free to ignore